server/src/main/java/org/opensearch/common/compress/CompressionHelper.java

```diff
@@ -0,0 +1,99 @@
+import org.opensearch.common.io.stream.Writeable;
+import org.opensearch.transport.BytesTransportRequest;
+
+import java.io.IOException;
+
+/**
+ * A helper class to utilize the compressed stream.
+ */
+public class CompressionHelper {
+    private static final Logger logger = LogManager.getLogger(CompressionHelper.class);
+
+    /**
+     * Serialize the given cluster state or diff. It'll always use compression before writing on a newly created output
+     * stream.
+     *
+     * @param writer             Object which is going to write the content
+     * @param node               version of cluster node
+     * @param isFullClusterState flag used at receiver end to make intelligent decisions. For example, ClusterState
+     *                           assumes full state of diff of the states based on this flag.
+     * @return reference to serialized bytes
+     * @throws IOException if writing on the compressed stream is failed.
+     */
+    public static BytesReference serializeClusterState(Writeable writer, DiscoveryNode node, boolean isFullClusterState)
+        throws IOException {
+        try {
+            final BytesStreamOutput bStream = new BytesStreamOutput();
+            try (StreamOutput stream = new OutputStreamStreamOutput(CompressorFactory.COMPRESSOR.threadLocalOutputStream(bStream))) {
+                stream.setVersion(node.getVersion());
+                stream.writeBoolean(isFullClusterState);
+                writer.writeTo(stream);
+            }
+            final BytesReference serializedByteRef = bStream.bytes();
+            logger.trace("serialized writable object for node version [{}] with size [{}]", node.getVersion(), serializedByteRef.length());
+            return serializedByteRef;
+        } catch (Exception e) {
+            logger.warn(() -> new ParameterizedMessage("failed to serialize cluster state during validateJoin" + " {}", node), e);
+            throw e;
+        }
+    }
+
+    public static StreamInput decompressClusterState(BytesTransportRequest request, NamedWriteableRegistry namedWriteableRegistry)
+        throws IOException {
+        final Compressor compressor = CompressorFactory.compressor(request.bytes());
+        StreamInput in = request.bytes().streamInput();
+        if (compressor != null) {
+            in = new InputStreamStreamInput(compressor.threadLocalInputStream(in));
+        }
+        in = new NamedWriteableAwareStreamInput(in, namedWriteableRegistry);
+        in.setVersion(request.version());
+        return in;
+    }
+
+    public static ClusterState deserializeFullClusterState(StreamInput in, DiscoveryNode localNode) throws IOException {
+        final ClusterState incomingState;
+        try (StreamInput input = in) {
+            incomingState = ClusterState.readFrom(input, localNode);
+        } catch (Exception e) {
+            logger.warn("unexpected error while deserializing an incoming cluster state", e);
+            throw e;
+        }
```

Comment:
whats the point catching and re-throwing the same exception, this might lead to logging redundant stack traces

Reply:
Agreed. Log-and-throw is an anti-pattern because the thing that ultimately handles this exception has the most context of what to do (and usually logs the exception anyway, leading to duplicate logging and stack traces).